File: frontend/types/feed.ts
/* Mirrors backend/schemas/feed.py */

import type { PublicUserSummary, Visibility } from "@/types/social";

export type FeedEventType = "workout" | "personal_record" | "routine_shared";

/* Which authors a page is drawn from */
export type FeedScope = "following" | "me" | "user";

/* Denormalised on the event, so a card renders without a follow-up request */
export interface WorkoutPayload {
  duration_seconds?: number;
  exercise_count?: number;
  set_count?: number;
  total_reps?: number;
  total_volume_kg?: number;
  total_distance_km?: number;
}

export interface RecordPayload {
  exercise?: string;
  kind?: string;
  weight_kg?: number;
  reps?: number;
  one_rm?: number;
  previous_one_rm?: number;
  improvement_kg?: number;
  session_name?: string;
}

export interface RoutinePayload {
  routine_name?: string;
  exercise_count?: number;
}

export type FeedPayload = WorkoutPayload & RecordPayload & RoutinePayload;

export interface FeedEventItem {
  id: string;
  event_type: FeedEventType;
  visibility: Visibility;
  author: PublicUserSummary;
  subject_type: string | null;
  subject_id: string | null;
  title: string;
  payload: FeedPayload;
  like_count: number;
  comment_count: number;
  liked_by_me: boolean;
  occurred_at: string | null;
}

export interface FeedPage {
  items: FeedEventItem[];
  /* Null on the last page, which is what the client checks rather than
     comparing the item count against the limit */
  next_cursor: string | null;
}

export interface FeedLikeResult {
  event_id: string;
  liked_by_me: boolean;
  like_count: number;
}

export interface FeedCommentItem {
  id: string;
  event_id: string;
  author: PublicUserSummary;
  body: string;
  created_at: string | null;
  /* True for the comment's author and for the owner of the post */
  can_delete: boolean;
}

File: backend/schemas/feed.py
from datetime import datetime
from typing import Any, Dict, List, Literal, Optional
from pydantic import BaseModel, Field, field_validator
from core.feed import FeedError, normalize_comment
from schemas.social import PublicUserSummary, Visibility
from typing import Dict, Any

EventType = Literal["workout", "personal_record", "routine_shared"]

# Which authors a feed page is drawn from
FeedScope = Literal["following", "me", "user"]


class FeedEventItem(BaseModel):
    """One card in the feed, self-contained so a page needs no follow-up reads"""

    id: str
    event_type: EventType
    visibility: Visibility
    author: PublicUserSummary

    subject_type: Optional[str] = None
    subject_id: Optional[str] = None

    title: str
    payload: Dict[str, Any] = Field(default_factory=dict)

    like_count: int = 0
    comment_count: int = 0

    # Resolved for the caller, so the like button renders without a second call
    liked_by_me: bool = False

    occurred_at: Optional[datetime] = None

    class Config:
        from_attributes = True


class FeedPage(BaseModel):
    """
    A keyset page of events

    `next_cursor` is null when the page is the last one, which is what the
    client checks rather than comparing the item count to the limit
    """

    items: List[FeedEventItem] = []
    next_cursor: Optional[str] = None


class FeedLikeResult(BaseModel):
    """The state a like button should show after a toggle"""

    event_id: str
    liked_by_me: bool
    like_count: int


class FeedCommentCreate(BaseModel):
    body: str

    @field_validator("body")
    @classmethod
    def check_body(cls, value: str) -> str:
        try:
            return normalize_comment(value)
        except FeedError as exc:
            raise ValueError(str(exc)) from exc


class FeedCommentItem(BaseModel):
    id: str
    event_id: str
    author: PublicUserSummary
    body: str
    created_at: Optional[datetime] = None
    # True when the caller may delete it, as the comment's author or the
    # event's owner. Moderating your own post is the minimum a feed needs
    can_delete: bool = False

    class Config:
        from_attributes = True


class FeedShareRequest(BaseModel):
    event_type: str
    subject_id: str
    title: str
    payload: Dict[str, Any] = Field(default_factory=dict)
    visibility: str = "followers"
